File: lc_soln/0110_balanced_binary_tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root)[0]

    def dfs(self, root):
        if root is None:
            return True, 0
        left_balanced, left_height = self.dfs(root.left)
        right_balanced, right_height = self.dfs(root.right)
        return left_balanced and right_balanced and abs(
            left_height - right_height
        ) < 2, 1 + max(left_height, right_height)


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0110_balanced_binary_tree/solution.rs
struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::dfs(&root).0
    }

    fn dfs(root: &Option<Rc<RefCell<TreeNode>>>) -> (bool, i32) {
        match root {
            None => (true, 0),
            Some(node) => {
                let node = node.borrow();
                let left = Self::dfs(&node.left);
                let right = Self::dfs(&node.right);
                let balanced = left.0 && right.0 && (left.1 - right.1).abs() < 2;
                (balanced, 1 + left.1.max(right.1))
            }
        }
    }
}

fn main() {}
